### ISSUE TITLE
Ensure add-on release origin is authentic

### ISSUE THREAD

**Issue by feerrenrut** [nvaccess/addon-datastore#41]
PR #35 demonstrates a potential issue with automatic approval of PRs. Data in a prior submission can be changed, pass verification, and potentially auto-merged. In this case the author and license are changed. It would also be possible to change the download URL and SHA.

Some ways to address this:
- Only allow automatic approval for changes to certain fields, (EG description). An add-on release is expected to be immutable, once submitted, users may have it installed. To fix an issue, a new version should be released.
- Don't allow file deletion without manual approval (requiring explanation).
- Introduce a way for authors to digitally sign add-on releases, so their origin can be confirmed.

**Comment by nvdaes** [comment]
For the following:

< - Introduce a way for authors to digitally sign add-on releases, so their origin can be confirmed.

PR #37 may provide this, since the publisher in metadata is set from the GitHub username that creates the issue that creates the PR to automerge.

**Comment by feerrenrut** [comment]
No, this isn't enough. To prevent hijacking of the add-on there needs to be a way to check that the person (or people) that prepared the add-on release haven't changed.

I've considered whether it could be mandated the addon release is tied to a github user / org, but there are some problems I don't like the inflexibility for the author. It also brings up problems when a new maintainer takes over.

Instead, if the addon package is signed, we know that the author of that version had access to the private key. We would have to store the public key of the author, or perhaps a URL and hash of it with each release.

**Comment by nvdaes** [comment]
Maybe good to test with public keys to know the possible difficulties and a way to document the procedure so that it"s easy and comfortable, and well understood why to do things like that.

Enviado desde mi iPhone

**Comment by feerrenrut** [comment]
Yes, this will require some investigation. We don't intend to require it early on. But it is worth being aware.

**Comment by nvdaes** [comment]
Personally, I'll experiment signing releases as explained at
https://wiki.debian.org/Creating%20signed%20GitHub%20releases

This may not be the final approach for the store, but I'll try to comment if I find special difficulties in case this can be useful.

**Comment by nvdaes** [comment]
I'm using the procedure described i the [GitHub help to create a new GPG](https://docs.github.com/en/authentication/managing-commit-signature-verification/generating-a-new-gpg-key) and related documents (options by default).
I installed Cleopatra, which has a gui for Windows, but seems not to be very accessible for me with NVDA since I cannot see each certificate.
Imo the process is not too difficult, but my advice would be to include this requirement as soon as the store is open for general testing, or alternatively make the signature process optional, not required, showing signed add-ons as verified and not signed add-ons without the verified state.
Otherwise, imo some people may feel this as something tricky and may not be motivated to share add-ons in the store. At least, if this is included ASAP, people maybe less frustrated with further requirements that may not be well understood, and unless this store has a great advantage over other existing procedures/websites to share add-ons, this probably will be an fforz by NV Access and testers that won't be very used by add-on authors.
Anyway people may prefer to share add-ons in the most easy way, but requiring this in a later stage may cause people to feel bad is this is not optional.

**Comment by feerrenrut** [comment]
I don't want to hold back that the store trying to add this requirement now. Given this problem has not previously been solved, I think we'll go on with the current approach. We may introduce it as an optional feature later.

**Comment by nvdaes** [comment]
Ok, anyway I will start signing add-on packages with detached-sign, attaching the .sig file to releases, since it is a good idea. When this optional feature is ready, I would like to test or use it, for example, we can provide an URL where the .sig file can be found, and public key maybe stored by you or shared in releases as an .asc file. This is very interesting as an optional feature.

Enviado desde mi iPhone

**Comment by nvdaes** [comment]
Hi again:

After investigation, I'll probably sign add-ons with GPG using GitHub Actions. I've done it just with my manualRelease.yaml workflow, which requires to provide the add-on version and if it's a prerelease. I've stored my private key and passphrase as repo secrets, and stored my public key in a publicKey.asc file in the rot of the project.
I'll share this on add-ons mailing list in case it's helpful:

- [manualRelease workflow](https://github.com/nvdaes/clipContentsDesigner/blob/main/.github/workflows/manualRelease.yaml)
- [import gpg action on GitHub marketplace](https://github.com/marketplace/actions/import-gpg)
- [release for testing: clipContentsDesigner-testGpg16.0.0-dev](https://github.com/nvdaes/clipContentsDesigner/releases/tag/testGpg16.0.0-dev)

**Comment by nvdaes** [comment]
Now Github has Artifact Attestations in beta.

**Comment by seanbudd** [comment]
I think we could do this by adding the following requirements:

1. The add-on is built by GitHub actions
1. The add-on is released via GitHub actions, and the release is [immutable](https://docs.github.com/en/code-security/concepts/supply-chain-security/immutable-releases)
1. We only pull add-ons from immutable releases built by GitHub actions

Implementation wise, we would need to add this support to the add-on template.

In conjunction with #9954, the add-on store submission experience could be greatly simplified. You could register your repo once off, and then every GitHub release could get submitted to the store automatically.

This could be controversial to lock developers into a platform, but most add-ons on GitHub already via the add-on template.
I think it would be a major boost to the security and trust of add-ons submitted to the store.

Thoughts @nvdaes @CyrilleB79 @SaschaCowley @michaelDCurran  ?

**Comment by nvdaes** [comment]
I like this idea, and if it's implemented I'll use it.
I suggest to make this optional, so that people wishing to use different platforms can still submit add-ons to the store. But in the store dialog, perhaps in the details panel (since there are many columns in the add-ons list), we may reflect that an add-on has been submitted via this proposed procedure.
I was thinking to work in the implementation of the triaged RSS featureproposed in nvaccess/addonstore-api#4

I started investigating this with help of Copilot. But if this is considered more important, unless @CyrilleB79 or someone works on this, once implementation details are provided, I may help with this issue.

**Comment by seanbudd** [comment]
>I suggest to make this optional,

Do you mean registering a repo for automatic release submissions, or do you mean forcing add-ons to be immutable github releases?
Because immutable releases is the heart of this issue - the automatic submissions for a repo is just a separate idea for further down the line - that only really becomes possible if we enforce github for submissions.

**Comment by nvdaes** [comment]
@seanbudd wrote:

> Do you mean registering a repo for automatic release submissions, or do you mean forcing add-ons to be immutable github releases?

I mean register add-ons for automatic submissions. NVDA checks sha256, so in fact it makes sense that releases are inmutable.

**Comment by seanbudd** [comment]
Forcing publishers who want to submit to the store to have to use GitHub Actions to build their add-ons is a big change though. That is the focus of the comment I raised and this ticket more broadly

**Comment by nvdaes** [comment]
@seanbudd wrote:

> Forcing publisher who want to submit to the store to have to use GitHub Actions to build their add-ons is a big change though. That is the focus of the comment I raised and this ticket more broadly

Then perhaps is better making this just optional, since may be people who don't wish to use this approach.

**Comment by seanbudd** [comment]
That's not the original proposal here though - it's really all or nothing. There's not much point to this if it is optional.

**Comment by seanbudd** [comment]
What I am proposing in https://github.com/nvaccess/addon-datastore/issues/41#issuecomment-4839356510 is a set of rules for all add-on submissions to make sure that anyone can verify the build process and source code of an add-on file. This means when someone reviews an add-ons source code in github, it should be the actual source of truth for the add-on we are accepting. 

Forget the registration stuff, that's only if we go down this path and complete https://github.com/nvaccess/addon-datastore/issues/9954

**Comment by seanbudd** [comment]
We could make an add-on being "authentic" as optional, but I wouldn't want to maintain two different submission systems, and making it optional makes messaging around security difficult.

**Comment by nvdaes** [comment]
Your arguments make sense. Though I'm concerned about people not wishing to use this proposed procedure. But I vote in favour of implementing this, and I'm willing to contribute to add support for this is required.

**Comment by CyrilleB79** [comment]
For reference, there are 437 stable add-ons in the Store (including incompatible ones). Only 11 of them are not GitHub releases as shown on [this download stats page](https://cyrilleb79.github.io/NVDAAddonsStats/).

Other nice side effect: if all the add-ons are GitHub releases we can retrieve download stats as done on the linked page.

Regarding acceptability of these changes, I'm not sure at all. But I have no other solution than asking people more widely, e.g. on the add-on lmailing list and / or people having published the 11 add-ons which are not GitHub release.

**Comment by SaschaCowley** [comment]
@michaelDCurran and I are happy with the proposal in https://github.com/nvaccess/addon-datastore/issues/41#issuecomment-4839356510.

**Comment by seanbudd** [comment]
Great - I will send an email out to the add-ons community later to gauge feedback on using GitHub only, and immutable releases, for the add-on store.
I think this is low priority for now, but if we have time to rework the add-on store, this and #9954 would be next for us to streamline the system.
